### ISSUE TITLE
`jl.println` and `jl.display` do not work in Jupyter notebook

### ISSUE THREAD

**Issue by ShuhuaGao** [JuliaPy/PythonCall.jl#232]
The following test was made in a .ipynb notebook in VS code.
![Snipaste_2022-10-17_17-22-36](https://user-images.githubusercontent.com/20141984/196140852-624be2bd-35cd-486d-ad01-a3888cb5b7df.png)
We see that `println` has no output at all, while `display` leads to an error.

- Julia version 1.8.2
- Python version 3.9.12
- juliacall   0.9.7

**Comment by cjdoris** [comment]
If you do `%load_ext juliacall.ipython` first then it will work. I had never tried it from VSCode before, but am happy to see that it does!

https://cjdoris.github.io/PythonCall.jl/dev/compat/#IPython

(It's only in the dev version of the docs because it appears the docs have not built for the last few releases, that's a separate issue.)

**Comment by ShuhuaGao** [comment]
Thanks. I tested that `%load_ext juliacall.ipython` in VS Code notebook. It sometimes worked, and the behavior was weired.
- Work
![Snipaste_2022-10-18_10-00-38](https://user-images.githubusercontent.com/20141984/196317959-813bfb51-d78c-40bd-aa38-d1a50236c396.png)

- Not work. If I insert a `jl.versioninfo()` at the beginning, then `println` gives no output.
![Snipaste_2022-10-18_09-59-26](https://user-images.githubusercontent.com/20141984/196317972-85fbb58c-cb0d-43b4-889f-3b72949abc6b.png)

However, `jl.VERSION` does not have that side effect.

![Snipaste_2022-10-18_10-00-38](https://user-images.githubusercontent.com/20141984/196318370-05d09d71-3813-4bce-a5c5-4c8b37d2683a.png)

**Comment by cjdoris** [comment]
Turns out what I had implemented was pretty hacky. It's much better now, I've made a release, it certainly seems to be much better in VSCode now.

**Comment by ShuhuaGao** [comment]
That's great.  Now the above issue of `versioninfo()` is fixed. See figure below.
![Snipaste_2022-10-19_15-01-20](https://user-images.githubusercontent.com/20141984/196619962-6300b933-1cc0-4a6a-a198-01fdf4e707d6.png)

However, if I mix up `println` and `display` (or even `print` in Python), the output order seems corrupted.
![2](https://user-images.githubusercontent.com/20141984/196620394-87ccc839-e078-4299-ab58-dc75f6fcdf95.png)
![3](https://user-images.githubusercontent.com/20141984/196620403-a09d472d-0a86-45eb-a802-89ab74120d6c.png)

**Comment by cjdoris** [comment]
Yeah, that's because in Julia you can't (reliably) replace stdout in Julia, you can only redirect it, and that needs to happen asynchronously, so anything printed to stdout from Julia may appear out of order with things printed from Python. `display` should always appear in order though.

**Comment by ShuhuaGao** [comment]
> `display` should always appear in order though.
Yes, it is true in my test.

Besides, `flush` after `println` may be a workaround to enforce the print in order.
![Snipaste_2022-10-21_13-31-57](https://user-images.githubusercontent.com/20141984/197120042-a9949a3b-f246-45c8-a47f-b93ba54e06d4.png)

**Comment by MilesCranmer** [comment]
I threw this in PySR and it seems to work (modified from tqdm):

```python
try:
    get_ipython = sys.modules["IPython"].get_ipython

    if 'IPKernelApp' not in get_ipython().config:
        raise ImportError("console")

    print("Detected Jupyter notebook. Loading juliacall extension.")

    get_ipython().run_line_magic("load_ext", "juliacall")
except Exception:
    pass
```

**Comment by cjdoris** [comment]
That seems reasonable - if we also provide a mechanism (env var) to turn this behaviour off.

**Comment by MilesCranmer** [comment]
Great minds think alike! 😄 Here's the code I ended up releasing in PySR:

```python
# Next, automatically load the juliacall extension if we're in a Jupyter notebook
autoload_extensions = os.environ.get("PYSR_AUTOLOAD_EXTENSIONS", "yes")
if autoload_extensions in {"yes", ""} and jl_version >= (1, 9, 0):
    try:
        get_ipython = sys.modules["IPython"].get_ipython

        if "IPKernelApp" not in get_ipython().config:
            raise ImportError("console")

        print(
            "Detected Jupyter notebook. Loading juliacall extension. Set `PYSR_AUTOLOAD_EXTENSIONS=no` to disable."
        )

        # TODO: Turn this off if juliacall does this automatically
        get_ipython().run_line_magic("load_ext", "juliacall")
    except Exception:
        pass
elif autoload_extensions not in {"no", "yes", ""}:
    warnings.warn(
        "PYSR_AUTOLOAD_EXTENSIONS environment variable is set to something other than 'yes' or 'no' or ''."
    )
```

Want me to PR to juliacall? I think it is nice to have printing work automatically.

**Comment by cjdoris** [comment]
Yeah go for it. Please add a setting so this feature can be disabled.
